HW2/P3/mandelbrot.pyx

```diff
@@ -42,6 +58,134 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
+        AVX.float8 tmp1, tmp2, tmp3
+        np.float32_t [:, :] in_coords_r, in_coords_z
+        float new_out_counts[8]
+
+    # Separate the in_coords into real and imaginary parts
+    in_coords_r = np.real(in_coords)
+    in_coords_z = np.imag(in_coords)
+
+    # Get ourselves an AVX.float8 value which just contains '4' for the magnitude of z
+    z_mag_cutoff = AVX.float_to_float8(4)
+
+    # And also get ourselves something that just stores all 1's for incrementing
+    all_ones = AVX.float_to_float8(1)
+
+    assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
+    assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
+    assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
+
+    with nogil:
+        for i in prange(in_coords.shape[0], schedule='static', chunksize=1, num_threads=n_threads):
+            # We do a step size of 8 because we are going to handle 8 at a time
+            # The assert above guarantees that we will not miss any items in doing so
+            for j in range(0, in_coords.shape[1], 8):
+
+                # Construct our AVX values - gotta be a better way to do this
+                c_r_vals = AVX.make_float8(in_coords_r[i, j + 7],
+                        in_coords_r[i, j + 6],
+                        in_coords_r[i, j + 5],
+                        in_coords_r[i, j + 4], 
+                        in_coords_r[i, j + 3],
+                        in_coords_r[i, j + 2],
+                        in_coords_r[i, j + 1],
+                        in_coords_r[i, j])
+
+                c_z_vals = AVX.make_float8(in_coords_z[i, j + 7],
+                        in_coords_z[i, j + 6],
+                        in_coords_z[i, j + 5],
+                        in_coords_z[i, j + 4], 
+                        in_coords_z[i, j + 3],
+                        in_coords_z[i, j + 2],
+                        in_coords_z[i, j + 1],
+                        in_coords_z[i, j])
+
+                # Both the real and imaginary parts of z start as 0
+                z_r_vals = AVX.float_to_float8(0)
+                z_z_vals = AVX.float_to_float8(0)
+
+                # And also count the number of iterations for each AVX value
+                iter_counter = AVX.float_to_float8(0)
+
+                for iter in range(max_iterations):
+                    # For an imaginary number z = x + iy, |z|^2 = x^2 + y^2
+                    # Calculate the magnitudes of each complex number in our AVX representation
+                    z_mag = AVX.fmadd(z_r_vals, z_r_vals, AVX.mul(z_z_vals, z_z_vals))
+
+                    # Now get a mask that tells us which of these are less than 4
+                    z_mag_small = AVX.less_than(z_mag, z_mag_cutoff)
+
+                    # If nothing needs to be updated, we are done
+                    if (AVX.signs(z_mag_small) == 0):
+                        break
+
```

Comment:
Nice, this makes more sense.  I had my counts to output step within this if statement, so I also had to say that if I reached the maximum iteration to write that in too.  This is smarter, just having a break and sending the outputs to memory after your for loop.